File: problem_026.py
#!/usr/bin/env python

print("Project Euler - Problem 26")
print("Find the value of d < 1000 for which 1/d contains the longest recurring cycle.\n")

longest = [0, 0]
debug = False

for d in range(2, 1000):
    answer = [1]
    i = 0
    repeat = 1
    repeatLength = -1

    # Perform long division
    while len(answer) < 1500:
        tmp = answer[i]/d
        tmp2 = tmp*d
        answer[i] -= tmp2
        answer.append(0)
        if answer[i] < d:
            answer[i+1] += answer[i]*10
            answer[i] = 0
        answer[i] = tmp

        # Keep track if repeating digits.
        if len(answer) > 3:
            if answer[i] == answer[repeat]:
                repeat += 1
            else:
                repeat = 1
        if repeat >= 10:
            repeatLength = i-repeat+1
            if repeatLength > longest[0]:
                longest[0] = repeatLength
                longest[1] = d
            break

        i += 1

    # Print answers
    if debug == True:
        print("1/%d repeats after %d digits" % (d, repeatLength))
        print(answer)

print(longest[1])

File: problem_026.js
#!/usr/bin/node

console.log("Project Euler - Problem 26:");
console.log("Find the value of d < 1000 for which 1/d contains the longest recurring cycle.\n");

const limit = 1000;
let max_cycle_lenth = 0;
let max_cycle_seed = 0;

function find_cycle_length(a){
    // Go from end of array
    let last_index = a.length-1
    // Walk backwards looking for potential lengths
    for (var l=3; l<a.length/2; l++){
	// Walk backwards at potential lengths and stop if we hit non-matches
	let matched = true;
	for (var i=last_index; i>last_index-l; i--){
	    if (a[i] != a[i-l]){
		matched = false;
		break;
	    }
	}
	if (matched == true)
	    return l;
    }
    return -1;
}

for (var d=2; d<limit; d++){
    var digits = [];
    var n = 1;
    while (true){
	// Manual division
	while (n < d)
	    n *= 10
	let digit = Math.floor(n/d);
	digits.push(digit);
	n -= digit * d;

	// Look for non-repeating terminations
	if (n == 0 || digit == 0)
	    break;

	// Look for repetitions
	let cycle_length = find_cycle_length(digits);
	if (cycle_length != -1){
	    if (cycle_length > max_cycle_lenth){
		max_cycle_lenth = cycle_length;
		max_cycle_seed  = d;
	    }
	    break;
	}
    }
}

console.log(max_cycle_seed);

